app/scripts/lib/rpc-method-middleware/handlers/wallet-createSession/handler.ts

```diff
@@ -0,0 +1,250 @@
+    requestPermissionApprovalForOrigin: (
+      requestedPermissions: RequestedPermissions,
+    ) => Promise<{ permissions: RequestedPermissions }>;
+    sendMetrics: (
+      payload: MetaMetricsEventPayload,
+      options?: MetaMetricsEventOptions,
+    ) => void;
+    metamaskState: {
+      metaMetricsId: string;
+      permissionHistory: Record<string, unknown>;
+      accounts: Record<string, unknown>;
+    };
+    grantPermissions: (
+      ...args: Parameters<AbstractPermissionController['grantPermissions']>
+    ) => Record<string, ValidPermission<string, Caveat<string, Json>>>;
+  },
+) {
+  const { origin } = req;
+  if (!isPlainObject(req.params)) {
+    return end(invalidParams({ data: { request: req } }));
+  }
+  const { requiredScopes, optionalScopes, sessionProperties } = req.params;
+
+  if (sessionProperties && Object.keys(sessionProperties).length === 0) {
+    return end(new JsonRpcError(5302, 'Invalid sessionProperties requested'));
+  }
+
+  try {
+    const { normalizedRequiredScopes, normalizedOptionalScopes } =
+      validateAndNormalizeScopes(requiredScopes || {}, optionalScopes || {});
+
+    const supportedRequiredScopesObjects = getSupportedScopeObjects(
+      normalizedRequiredScopes,
+    );
+    const supportedOptionalScopesObjects = getSupportedScopeObjects(
+      normalizedOptionalScopes,
+    );
+
+    const existsNetworkClientForChainId = (chainId: Hex) => {
+      try {
+        hooks.findNetworkClientIdByChainId(chainId);
+        return true;
+      } catch (err) {
+        return false;
+      }
+    };
+
+    const { supportedScopes: supportedRequiredScopes } = bucketScopes(
+      supportedRequiredScopesObjects,
+      {
+        isChainIdSupported: existsNetworkClientForChainId,
+        isChainIdSupportable: () => false, // intended for future usage with eip3085 scopedProperties
+      },
+    );
+
+    const { supportedScopes: supportedOptionalScopes } = bucketScopes(
+      supportedOptionalScopesObjects,
+      {
+        isChainIdSupported: existsNetworkClientForChainId,
+        isChainIdSupportable: () => false, // intended for future usage with eip3085 scopedProperties
+      },
+    );
+
+    // Fetch EVM accounts from native wallet keyring
+    // These addresses are lowercased already
+    const existingEvmAddresses = hooks
+      .listAccounts()
+      .map((account) => account.address);
+    const supportedEthAccounts = getEthAccounts({
+      requiredScopes: supportedRequiredScopes,
+      optionalScopes: supportedOptionalScopes,
+    })
+      .map((address) => address.toLowerCase() as Hex)
+      .filter((address) => existingEvmAddresses.includes(address));
+
+    const requestedCaip25CaveatValue = {
+      requiredScopes: getInternalScopesObject(supportedRequiredScopes),
+      optionalScopes: getInternalScopesObject(supportedOptionalScopes),
+      isMultichainOrigin: true,
+    };
+
+    const requestedCaip25CaveatValueWithSupportedEthAccounts = setEthAccounts(
```

Comment:
Why do we have to perform this step? All of the accounts listed in `requestedCaip25CaveatValue` should _already_ be supported (because we filtered out the non-supported ones). Or is it that we are now making sure that the caveat value _only_ contains EVM accounts?

Reply:
Answering my own question, I think that is what's happening here. Also all of the addresses here are guaranteed to be lowercase.